### PR TITLE
[receiver/k8scluster] add support for observing resources for a specific namespace

### LINKED ISSUE #9401: k8sclusterreceiver: Fetch namespace scoped metrics from the Master API
**Is your feature request related to a problem? Please describe.**
The k8sclusterreceiver today is set to fetch metrics from the k8s master API at the cluster scope level. In cases where we'd like to deploy the otel-collector as a sidecar, there is no mechanism to fetch metrics scoped to the namespace level. E.g. The pod phases and pod restarts for the namespace the collector is deployed in as a sidecar.

Ideally, there is a sidecar deployment model for the collector where the k8sclusterreceiver and the kubeletstatsreceiver both fetch data only related to the pod they are deployed on. This will help immensely for developers who do not have access to leverage ClusterRoles and are restricted to a certain set of namespaces.

**Describe the solution you'd like**
The receiver should have a flag to restrict fetching data from the namespace it is deployed in. 

**Describe alternatives you've considered**
I tried setting a namespace scoped Role for the serviceAccount on the k8sclusterreceiver. It unfortunately still hits the `/api/v1/pods` endpoint and I'd like to use the `/api/v1/namespaces/<namespace-name>/pods` endpoint.

### PR DESCRIPTION
**Pull request by bacherfl**
#### Description

This PR extends the k8scluster receiver with an option to limit the observed resources to a specific namespace.

#### Link to tracking issue
Fixes #9401 

#### Testing

added unit and e2e tests

#### Documentation

Added section about how to make use of Roles and RoleBindings instead of ClusterRoles and ClusterRoleBindings

### CODE REVIEW

File: receiver/k8sclusterreceiver/config.go
Comment (ChrsMark) [inline receiver/k8sclusterreceiver/config.go:46]: I wonder if this effect should also be logged and/or be reflected to the affected resources' settings.

It would be weird to have the following:

```yaml
  k8s_cluster:
    auth_type: kubeConfig
    node_conditions_to_report: [Ready, MemoryPressure]
    namespace: staging
``` 

So would that make sense to ensure that affected metrics and settings are explicitly disabled in case the `namespace` is defined?

Maybe that's an overkill though so I would rely on code-owner's opinions here :)

**Review by povilasv — APPROVED** [review]

File: receiver/k8sclusterreceiver/config.go
Comment (bacherfl) [reply receiver/k8sclusterreceiver/config.go:46]: That's a good point, thanks @ChrsMark - I think it does make sense to check for this during the config validation, I'll add a check for that :)

**Review by ChrsMark — APPROVED** [review]

File: receiver/k8sclusterreceiver/config.go
Comment (TylerHelmuth) [inline receiver/k8sclusterreceiver/config.go:45]: I don't like recommending the receiver be used in a sidecar.  Even if it is only scoped to getting namespace-specific metrics, if it is run in multiple sidecars within that namespace you'll produce duplicate metrics.

Comment (bacherfl) [reply receiver/k8sclusterreceiver/config.go]: Good point, I have removed that part from the comment

**Review by TylerHelmuth — APPROVED** [review]

**Comment by ChrsMark** [comment]
This has 2/3 code-owners' approval. Marking as `ready-to-merge`.

**Comment by bacherfl** [comment]
It seems like after adapting to the recent changes on `main` the e2e tests are failing now - i'm currently looking into this and should have this resolved soon

**Comment by ChrsMark** [comment]
I see e2e tests are failing on `main` anyways: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11701936581/job/32589392145

I wonder if https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36114 introduced some flakiness.

**Comment by bacherfl** [comment]
> I see e2e tests are failing on `main` anyways: https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11701936581/job/32589392145
> 
> I wonder if #36114 introduced some flakiness.

yes i think part of the flakiness might be due to the addition of the cronjob object which causes multiple pods to spin up during the duration of the test, leading to a varying number of metrics received in the sink

**Comment by ChrsMark** [comment]
> yes i think part of the flakiness might be due to the addition of the cronjob object which causes multiple pods to spin up during the duration of the test, leading to a varying number of metrics received in the sink

Sounds plausible, we should restrict the number of active jobs/pods to 1 to avoid having scheduling/timing related flakiness: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36114/files#diff-c08d5cc648ae8a9f364187a6dbf9a8b1e01070249b7a8669537009ccc56e701fR147.

I will try to send a PR for this.

File: receiver/k8sclusterreceiver/e2e_test.go
Comment (ChrsMark) [inline receiver/k8sclusterreceiver/e2e_test.go:293]: Is this needed for a specific reason?

File: receiver/k8sclusterreceiver/e2e_test.go
Comment (ChrsMark) [inline receiver/k8sclusterreceiver/e2e_test.go:72]: Is this change needed here? In that case, I wonder if we should do any refactoring to the existing tests in a separate PR.

**Review by ChrsMark — COMMENTED** [review]
Let's wait for https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/36235 to get merged and re-base this one. Meanwhile I'm a bit worried about the changes to the existing tests.

**Comment by bacherfl** [comment]
> Let's wait for #36235 to get merged and re-base this one. Meanwhile I'm a bit worried about the changes to the existing tests.

agree, i did some changes in the meantime and got the tests working on this PR, but some changes may have been unneccessary - I will try out everything after rebasing and keep the changes as minimal as possible

**Comment by bacherfl** [comment]
Alright this should now be ready again - @ChrsMark I reverted the changes I made earlier in the `waitForData` function as it turned out with the recent fix being merged they were not required anymore

File: receiver/k8sclusterreceiver/e2e_test.go
Comment (ChrsMark) [inline receiver/k8sclusterreceiver/e2e_test.go:71]: nit: I would suggest we don't remove this since it's useful to just uncomment that line in order to update the golden files. Maybe adding a comment to describe its usage would also help. Same applies for the newly added test.

**Review by ChrsMark — APPROVED** [review]
LGTM. Left a nit comment. Thank's @bacherfl!

(it will need a [`make generate`](https://github.com/open-telemetry/opentelemetry-collector-contrib/actions/runs/11737802610/job/32703905696?pr=35727#step:16:1503) to make the CI happy)

File: receiver/k8sclusterreceiver/e2e_test.go
Comment (bacherfl) [reply receiver/k8sclusterreceiver/e2e_test.go:71]: good catch, thanks - I re-added the comment

**Comment by ChrsMark** [comment]
@TylerHelmuth that should be good to go

**Pull request merged**
